plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/MavenFacetImpl.java

```diff
@@ -0,0 +1,413 @@
+  @Override
+  public VersionPolicy getVersionPolicy() {
+    return versionPolicy;
+  }
+
+  @Nonnull
+  @Override
+  public ChecksumPolicy getChecksumPolicy() {
+    return checksumPolicy;
+  }
+
+  @Nullable
+  @Override
+  public Payload get(final MavenPath path) throws IOException {
+    try (StorageTx tx = getStorage().openTx()) {
+      final OrientVertex asset;
+      if (path.getCoordinates() != null) {
+        final OrientVertex component = findComponent(tx, path.getCoordinates(), tx.getBucket());
+        if (component == null) {
+          return null;
+        }
+        asset = selectComponentAsset(tx, component, path);
+      }
+      else {
+        asset = findAsset(tx, path, tx.getBucket());
+      }
+      if (asset == null) {
+        return null;
+      }
+      final BlobRef blobRef = getBlobRef(path, asset);
+      final Blob blob = tx.getBlob(blobRef);
+      checkState(blob != null, "asset of component with at path %s refers to missing blob %s", path.getPath(),
+          blobRef);
+      final String contentType = asset.getProperty(StorageFacet.P_CONTENT_TYPE);
+      return new BlobPayload(blob, contentType);
+    }
+  }
+
+  @Override
+  public void put(final MavenPath path, final Payload content)
+      throws IOException, InvalidContentException
+  {
+    try (StorageTx tx = getStorage().openTx()) {
+      OrientVertex component = null;
+      OrientVertex asset;
+      if (path.getCoordinates() != null) {
+        component = findComponent(tx, path.getCoordinates(), tx.getBucket());
+        if (component == null) {
+          component = tx.createComponent(tx.getBucket());
+
+          // Set normalized properties: format, group, and name
+          component.setProperty(StorageFacet.P_FORMAT, getRepository().getFormat().getValue());
+          component.setProperty(StorageFacet.P_PATH, path.getPath());
+          component.setProperty(StorageFacet.P_GROUP, path.getCoordinates().getGroupId());
+          component.setProperty(StorageFacet.P_NAME, path.getCoordinates().getArtifactId());
+          component.setProperty(StorageFacet.P_VERSION, path.getCoordinates().getVersion());
+
+          final NestedAttributesMap componentAttributes = getFormatAttributes(tx, component);
+          componentAttributes.set(P_COMPONENT_KEY, getComponentKey(path.getCoordinates()));
+          componentAttributes.set(P_GROUP_ID, path.getCoordinates().getGroupId());
+          componentAttributes.set(P_ARTIFACT_ID, path.getCoordinates().getArtifactId());
+          componentAttributes.set(P_VERSION, path.getCoordinates().getVersion());
+          if (path.getCoordinates().isSnapshot()) {
+            componentAttributes.set(P_BASE_VERSION, path.getCoordinates().getBaseVersion());
+          }
+        }
+
+        asset = selectComponentAsset(tx, component, path);
+      }
+      else {
+        asset = findAsset(tx, path, tx.getBucket());
+      }
+      if (asset == null) {
+        asset = tx.createAsset(tx.getBucket());
```

Comment:
Same here, 'extract method' - keep the logic of the outer method a lot easier to see.